### PR TITLE
Add UserHasBeenDeactivated plugin hook

### PR DESCRIPTION
**Pull request by MatthewDorner**
#### Summary
Add a new Plugin hook, `UserHasBeenDeactivated` as specified by @mickmister here: https://github.com/mattermost/mattermost-plugin-gitlab/issues/287. Goes with the plugin PR: https://github.com/mattermost/mattermost-plugin-gitlab/pull/315.

#### Ticket Link
https://github.com/mattermost/mattermost-plugin-gitlab/issues/287

#### Release Note
```release-note
Added the Plugin hook "UserHasBeenDeactivated"
```

### CODE REVIEW

File: plugin/hooks.go
Comment (MatthewDorner) [inline plugin/hooks.go:277]: What should I put for the minimum server version?

**Comment by MatthewDorner** [comment]
Does the new hook need to be added to some documentation repos?

**Review by mickmister — APPROVED** [review]
Thanks for this @matthewdorner :+1: Functionality LGTM, just one non-blocking comment on test data

File: plugin/hooks.go
Comment (mickmister) [reply plugin/hooks.go:278]: @ashishbhate The context passed into hooks is the `plugin.Context` struct. In this case, I don't think the `plugin.Context` struct would be all that useful, but we can pass it for consistency and future-proofing. It also seems that many of the methods prefer `*model.User` rather than accepting a user ID

Comment (ashishbhate) [reply plugin/hooks.go:278]: @mickmister I'm 2/5 that we should add a `plugin.Context` and use `*model.User`. I'll defer to you though.

File: app/plugin_hooks_test.go
Comment (MatthewDorner) [reply app/plugin_hooks_test.go]: Done, although there are still instances of this in other files.

File: plugin/hooks.go
Comment (MatthewDorner) [reply plugin/hooks.go]: @ashishbhate @mickmister I added these changes

**Comment by MatthewDorner** [comment]
Hmm, didn't realize I could only request review from one person lol.

**Review by hanzei — APPROVED** [review]
LGTM :+1:

File: app/plugin_hooks_test.go
Comment (mickmister) [inline app/plugin_hooks_test.go:857]: About the test data, I was mainly saying that we shouldn't further copy it into new tests. This code is not part of this PR's scope so it shouldn't necessarily be edited here. In this sort of situation, it's the commit history of the individual lines that I would be worried about, but the history of these specific lines are not important, so there's not really a problem here. Thanks for cleaning this up :+1:

**Review by mickmister — DISMISSED** [review]
LGTM :+1:

**Review by ashishbhate — APPROVED** [review]

**Comment by hanzei** [comment]
@DHaussermann This PR should be testing alongside with https://github.com/mattermost/mattermost-plugin-gitlab/pull/315

**Comment by isacikgoz** [comment]
Hi @MatthewDorner, big thanks for your contribution! There hasn't been recent activity on this PR so I've marked it as inactive. Following our [inactive contribution process](https://developers.mattermost.com/contribute/getting-started/inactive-contributions/), the PR is now closed due to 30 days or more of inactivity.

However, if you're still interested working on the changes, let us know! Note that we [merged multiple repositories into one monorepo](https://github.com/mattermost/mattermost-server/issues/22420), which may require you to resubmit the changes. 

If you have any questions, don't hesitate to let us know, we're happy to help.

**Comment by hanzei** [comment]
@isacikgoz Can you please elaborate on why the PR was closed? It was blocked by the missing review of @DHaussermann, not because an action from the community member was missing.

**Comment by DHaussermann** [comment]
@hanzei and @isacikgoz we could re-open this. 

That said I don't know of an option for functional testing. We might need to implement the use of the hook in the demo plugin.

**Comment by hanzei** [comment]
@DHaussermann @MatthewDorner provided an implemenation of the hook via https://github.com/mattermost/mattermost-plugin-gitlab/pull/315. Can you use that for testing?

**Comment by isacikgoz** [comment]
@hanzei Sorry for the late reply, I was OOO for a while. I closed many PRs to cleanup some stale work. For this one I may missed the context and any time we can re-open this if necessary. Feel free to do so.

**Review by hanzei — APPROVED** [review]

**Comment by hanzei** [comment]
I've submitted a PR to the plugin that makes use of the new hook.

@DHaussermann I've tested the above-linked PR and can confirm that this PR works as expected. Are we good in terms of QA testing then?

**Comment by hanzei** [comment]
@MatthewDorner Would you mind If I also add a `UserHasBeenActivated` hook?

**Review by mickmister — APPROVED** [review]

**Comment by mickmister** [comment]
@hanzei LGTM. Also in agreement this is good to merge

**Comment by mickmister** [comment]
@hanzei LGTM. Also in agreement this is good to merge @DHaussermann

**Comment by hanzei** [comment]
QA deferred by the above comment. cc @DHaussermann

**Comment by DHaussermann** [comment]
Understood. 
Thanks @hanzei!

**Pull request merged**
